irohad/network/block_loader.hpp

```diff
@@ -7,13 +7,28 @@
 #define IROHA_BLOCK_LOADER_HPP
 
 #include <memory>
-#include <rxcpp/rx-observable-fwd.hpp>
+#include <variant>
 
 #include "interfaces/common_objects/types.hpp"
 #include "interfaces/iroha_internal/block.hpp"
 
 namespace iroha {
   namespace network {
+    class BlockReader {
+     public:
+      /**
+       * Try to read the next block. Returns monostate when the iteration is
+       * completed, std::string when an error occurred
+       */
+      virtual std::variant<
+          std::monostate,
+          std::shared_ptr<const shared_model::interface::Block>,
+          std::string>
+      read() = 0;
```

Comment:
I'd suggest to create empty `struct iteration_complete` instead of general std::monostate to indicate the meaning.
```suggestion
       * Try to read the next block. Returns monostate when the iteration is
       * completed, std::string when an error occurred
       */
      struct iteration_complete {};
      virtual std::variant<
          iteration_complete,
          std::shared_ptr<const shared_model::interface::Block>,
          std::string>
      read() = 0;
```